service/history/queues/queue_scheduled.go

```diff
@@ -0,0 +1,247 @@
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	p.logger.Info("", tag.LifeCycleStarting)
+	defer p.logger.Info("", tag.LifeCycleStarted)
+
+	p.queueBase.Start()
+
+	p.shutdownWG.Add(1)
+	go p.processEventLoop()
+
+	p.notify(time.Time{})
+}
+
+func (p *scheduledQueue) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	p.logger.Info("", tag.LifeCycleStopping)
+	defer p.logger.Info("", tag.LifeCycleStopped)
+
+	close(p.shutdownCh)
+	p.timerGate.Close()
+
+	if success := common.AwaitWaitGroup(&p.shutdownWG, time.Minute); !success {
+		p.logger.Warn("", tag.LifeCycleStopTimedout)
+	}
+
+	p.queueBase.Stop()
+}
+
+func (p *scheduledQueue) NotifyNewTasks(_ string, tasks []tasks.Task) {
+	if len(tasks) == 0 {
+		return
+	}
+
+	newTime := tasks[0].GetVisibilityTime()
+	for _, task := range tasks {
+		ts := task.GetVisibilityTime()
+		if ts.Before(newTime) {
+			newTime = ts
+		}
+	}
+
+	p.notify(newTime)
+}
+
+func (p *scheduledQueue) processEventLoop() {
+	defer p.shutdownWG.Done()
+
+	for {
+		select {
+		case <-p.shutdownCh:
+			return
+		case <-p.newTimerCh:
+			p.processNewTime()
+		case <-p.timerGate.FireChan():
+			p.processNewRange()
+		case <-p.pollTimer.C:
+			p.processPollTimer()
+		case <-p.checkpointTimer.C:
+			p.checkpoint()
+		}
+	}
+}
+
+func (p *scheduledQueue) notify(newTime time.Time) {
+	p.newTimeLock.Lock()
+	defer p.newTimeLock.Unlock()
+
+	if !p.newTime.IsZero() && !newTime.Before(p.newTime) {
+		return
+	}
+
+	p.newTime = newTime
+	select {
+	case p.newTimerCh <- struct{}{}:
+	default:
+	}
+}
+
+func (p *scheduledQueue) processNewTime() {
+	p.newTimeLock.Lock()
+	newTime := p.newTime
+	p.newTime = time.Time{}
+	p.newTimeLock.Unlock()
+
+	p.timerGate.Update(newTime)
+}
+
+func (p *scheduledQueue) processNewRange() {
+	p.queueBase.processNewRange()
+	p.lookAheadTask()
+}
+
+func (p *scheduledQueue) lookAheadTask() {
+	lookAheadMinTime := p.nonReadableScope.Range.InclusiveMin.FireTime
+	lookAheadMaxTime := lookAheadMinTime.Add(p.options.MaxPollInterval())
+
+	ctx, cancel := newQueueIOContext()
+	defer cancel()
+
+	request := &persistence.GetHistoryTasksRequest{
+		ShardID:             p.shard.GetShardID(),
+		TaskCategory:        tasks.CategoryTimer,
+		InclusiveMinTaskKey: tasks.NewKey(lookAheadMinTime, 0),
+		ExclusiveMaxTaskKey: tasks.NewKey(lookAheadMaxTime, 0),
+		BatchSize:           1,
+		NextPageToken:       nil,
+	}
+	response, err := p.shard.GetExecutionManager().GetHistoryTasks(ctx, request)
+	if err != nil {
+		// NOTE: the backoff is actually 2*TimerProcessorMaxTimeShift = 2s
+		p.timerGate.Update(lookAheadMinTime)
+		p.logger.Error("Failed to load look ahead task", tag.Error(err))
+		return
+	}
+
+	if len(response.Tasks) == 1 {
+		p.timerGate.Update(response.Tasks[0].GetKey().FireTime)
+	}
+
+	// no look ahead task, wait for max poll interval or new task notification
```

Comment:
do you need to reset the timer gate?

Reply:
Good catch! Performed a lookahead in after processing poll timer as well.

Reply:
Will add a test for this case.